### PR TITLE
Cache data file with IndexedDB

### PR DESCRIPTION
**Pull request by thomasballinger**
It seems like this works. I've deployed it to https://play-endless-sky.com/ and it downloads the first time normally (although no more download progress for now) and the second time skips the download.

To do this right, I think we want to hash the file or use the current git hash during the build process and store the that key along with the data in IndexedDB.

### CODE REVIEW

File: es2.html [hunk 1]
```diff
@@ -51,6 +51,10 @@
...
     
+    <script>
+      // this is idb-keyval https://www.npmjs.com/package/idb-keyval
+var idbKeyval=function(e){"use strict";class t{constructor(e="keyval-store",t="keyval"){this.storeName=t,this._dbp=new Promise((r,n)=>{const o=indexedDB.open(e,1);o.onerror=(()=>n(o.error)),o.onsuccess=(()=>r(o.result)),o.onupgradeneeded=(()=>{o.result.createObjectStore(t)})})}_withIDBStore(e,t){return this._dbp.then(r=>new Promise((n,o)=>{const s=r.transaction(this.storeName,e);s.oncomplete=(()=>n()),s.onabort=s.onerror=(()=>o(s.error)),t(s.objectStore(this.storeName))}))}}let r;function n(){return r||(r=new t),r}return e.Store=t,e.get=function(e,t=n()){let r;return t._withIDBStore("readonly",t=>{r=t.get(e)}).then(()=>r.result)},e.set=function(e,t,r=n()){return r._withIDBStore("readwrite",r=>{r.put(t,e)})},e.del=function(e,t=n()){return t._withIDBStore("readwrite",t=>{t.delete(e)})},e.clear=function(e=n()){return e._withIDBStore("readwrite",e=>{e.clear()})},e.keys=function(e=n()){const t=[];return e._withIDBStore("readonly",e=>{(e.openKeyCursor||e.openCursor).call(e).onsuccess=function(){this.result&&(t.push(this.result.key),this.result.continue())}}).then(()=>t)},e}({});
```
Comment (janisozaur) [inline es2.html]: Is there a specific reason this isn't formatted? Is this compiled version of https://github.com/jakearchibald/idb-keyval?

**Comment by janisozaur** [comment]
Looks good, but the key is required. Also - is there any action to take on mismatch, like evict old entries?

Regarding the key itself - I don't know how stable and reproducible is building the FS. For extra stability it would be best to hash the inputs, but that would require more changes to build system. Maybe taking a less strict route of taking just the hash of less strict isn't that bad, especially at this stage.

File: es2.html (hunk 1 above)
Comment (thomasballinger) [reply es2.html]: The minified version was what I found quickly, I'll move this to another file and generally clean up the JavaScript a bit.

**Comment by thomasballinger** [comment]
> Looks good, but the key is required. Also - is there any action to take on mismatch, like evict old entries?

Yeah we should do that, I'm thinking the key value store would just be this:
```
{
  data: <ArrayBuffer(86000000)> // data file contents
  version: "abcdef0123456789"  // hash 
}
```
and we check the version, so eviction is just replacing the one data value.

> Regarding the key itself - I don't know how stable and reproducible is building the FS. For extra stability it would be best to hash the inputs, but that would require more changes to build system. Maybe taking a less strict route of taking just the hash of less strict isn't that bad, especially at this stage.

Hm, I was thinking of using git commit sh, but that's also too strict. Are you saying hash of the output file endless-sky.data? Sounds good to me.

**Comment by thomasballinger** [comment]
I added hashing the file to the build process and writing that hash to a JavaScript file that can be read by the caching logic.

**Pull request merged**